azkaban-common/src/main/java/azkaban/executor/selector/ExecutorComparator.java

```diff
@@ -0,0 +1,286 @@
+ * the License.
+ */
+
+package azkaban.executor.selector;
+
+import java.util.Comparator;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+import azkaban.executor.Executor;
+import azkaban.executor.Statistics;
+
+public class ExecutorComparator extends CandidateComparator<Executor> {
+  private static Map<String, ComparatorCreator> comparatorCreatorRepository = null;
+
+  /**
+   * Gets the name list of all available comparators.
+   * @return the list of the names.
+   * */
+  public static Set<String> getAvailableComparatorNames(){
+    return comparatorCreatorRepository.keySet();
+  }
+
+  // factor comparator names
+  private static final String NUMOFASSIGNEDFLOW_COMPARATOR_NAME = "NumberOfAssignedFlowComparator";
+  private static final String MEMORY_COMPARATOR_NAME = "Memory";
+  private static final String PRIORITY_COMPARATOR_NAME = "Priority";
+  private static final String LSTDISPATCHED_COMPARATOR_NAME = "LastDispatched";
+  private static final String CPUUSAGE_COMPARATOR_NAME = "CpuUsage";
+
+  /**
+   * static initializer of the class.
+   * We will build the filter repository here.
+   * when a new comparator is added, please do remember to register it here.
+   * */
+  static {
+    comparatorCreatorRepository = new HashMap<String, ComparatorCreator>();
+
+    // register the creator for number of assigned flow comparator.
+    comparatorCreatorRepository.put(NUMOFASSIGNEDFLOW_COMPARATOR_NAME, new ComparatorCreator(){
+      public FactorComparator<Executor> create(int weight) { return getNumberOfAssignedFlowComparator(weight); }});
+
+    // register the creator for memory comparator.
+    comparatorCreatorRepository.put(MEMORY_COMPARATOR_NAME, new ComparatorCreator(){
+      public FactorComparator<Executor> create(int weight) { return getMemoryComparator(weight); }});
+
+    // register the creator for priority comparator.
+    comparatorCreatorRepository.put(PRIORITY_COMPARATOR_NAME, new ComparatorCreator(){
+      public FactorComparator<Executor> create(int weight) { return getPriorityComparator(weight); }});
+
+    // register the creator for last dispatched time comparator.
+    comparatorCreatorRepository.put(LSTDISPATCHED_COMPARATOR_NAME, new ComparatorCreator(){
+      public FactorComparator<Executor> create(int weight) { return getLstDispatchedTimeComparator(weight); }});
+
+    // register the creator for CPU Usage comparator.
+    comparatorCreatorRepository.put(CPUUSAGE_COMPARATOR_NAME, new ComparatorCreator(){
+      public FactorComparator<Executor> create(int weight) { return getCpuUsageComparator(weight); }});
+  }
+
+
+  /**
+   * constructor of the ExecutorComparator.
+   * @param comparatorList   the list of comparator, plus its weight information to be registered,
+   *  the parameter must be a not-empty and valid list object.
+   * */
+  public ExecutorComparator(Map<String,Integer> comparatorList) {
```

Comment:
total weight now doesn't have a cap (say you can have 100 or 1000 as weight) so I figured it doesn't make much sense to use double, int will be good enough.